### ISSUE TITLE
Extending KML the right way

### ISSUE THREAD

**Issue by sylvaneau** [samcragg/sharpkml#11]
Hi,

As you know, KML supports format extension in several ways, all explained in the KML specification document, section 6.6 :
http://portal.opengeospatial.org/files/?artifact_id=27810

One option is to extend existing elements using substitution (chapter 6.6.3.2), for instance a Folder exposes a kml:DocumentObjectExtensionGroup allowing us to add new children types to a folder. Like google did with its "gx" extension namespace.

What is the best way to translate this extension mecanism to SharpKML ?

It might be interisting to expose a `RegisterValidChild` method allowing us to register our extension. Or maybe an Attribute to follow what you did for ChildAttribute.

Thx again for your help

**Comment by samcragg** [comment]
Very good point. I think there are a couple of options.

The easiest may be to just let you subclass Folder etc, where you can then add any fields you want. However, I'm not sure how this would work when parsing the KML - how would the library know to use your version of the Folder rather than it's own.

A better option would be to add the extension properties to each class, as there are multiple extension points we can't just add it to the base `Element` class:

``` XML
<kml:Placemark ...>
<kml:ObjectSimpleExtensionGroup> ... </kml:ObjectSimpleExtensionGroup> [0..*]
<kml:AbstractFeatureSimpleExtensionGroup> ...
</kml:AbstractFeatureSimpleExtensionGroup> [0..*]
<kml:AbstractFeatureObjectExtensionGroup> ...
</kml:AbstractFeatureObjectExtensionGroup> [0..*]
<kml:PlacemarkSimpleExtensionGroup> ... </kml:PlacemarkSimpleExtensionGroup>
[0..*]
<kml:PlacemarkObjectExtensionGroup> ... </kml:PlacemarkObjectExtensionGroup>
[0..*]
</kml:Placemark>
```

I believe these should be actual properties in their respective types, as they have a unique name and because the classes are inherited in the same way as the KML ones we would correctly expose all of them.

What are your preferences? Is there a particular problem you are trying to solve that would give a good use case that we could focus on?

**Comment by sylvaneau** [comment]
I like your idea of subclassing any type, we just need to "register" again our type to replace the original one, it seems to be a simple solution.

I don't really see how you can use your second option, these properties are abstract, so you will never see an `AbstractFeatureObjectExtensionGroup` in KML. Basically the Google extension you've integrated to SharpKML uses this mecanism. We might need to find a common way to deal with the Google Extension and any other customer extension. Sorry if I didn't understand what you wanted to do.

Basically we want to extend a placemark or a folder just like Google did with its extension. For instance, we need to add an "editable" field to a feature :

```xml
<kml:Placemark>
  <ext:PlacemarkExtension>
    <ext:editable>false</ext:editable>
    <ext:linkedItems>
      <ext:item from="db">453135486</ext:item>
    </ext:linkedItems>
  </ext:PlacemarkExtension>
</kml:Placemark>
```
The `PlacemarkExtension` is even optionnal as KML standard allows us to directly extend the `Placemark` element but is just for human readability.

**Comment by samcragg** [comment]
That should work now in the latest version: https://www.nuget.org/packages/SharpKml.Core/3.3.0

I've added some documentation here as well, feel free to add a comment here (or a PR) if you want to expand it: https://github.com/samcragg/sharpkml/blob/master/docs/Extending.md

Thanks for the suggestion and hope it works as you were expecting 🙂

**Comment by sylvaneau** [comment]
Hi,

I think my idea of using inheritance wasn't a good one at all...
I ran into the following problem which, I think, can't be resolved using our system :-(

Let's say we want to Extend `Feature` and `Placemark`, I want to add an `Opacity` field to every feature and an `editable` field to `Placemark` only. The resulting KML would look like to this :

```xml
<kml:Placemark>
  <ext:FeatureExtension>
    <ext:opacity>1.0</ext:opacity>
  </ext:FeatureExtension>
  <ext:PlacemarkExtension>
    <ext:editable>false</ext:editable>
  </ext:PlacemarkExtension>
</kml:Placemark>
<kml:GroundOverlay>
  <ext:FeatureExtension>
    <ext:opacity>1.0</ext:opacity>
  </ext:FeatureExtension>
</kml:GroundOverlay>
```

`KmlFactory.Replace<TExisting, TNew>` works only with "final" types (i.e. `Placemark` or `GroundOverlay` but not `Feature`), so I can't register my ExtendedFeatureType... I could register `Placemark` and `GroundOverlay` but all the other features will miss the `Opacity` field...

So you were right, we should keep the KML model sealed and extend it another way...

**Comment by sylvaneau** [comment]
The second option might be the right one, lets take the same example.

We can add a collection of `AbstractFeatureObjectExtension` (probably inheriting from `Element`) to `Feature`.
Then we can create a `MyCompanyFeatureObjectExtension : AbstractFeatureObjectExtension` and register it to the KmlFactory as any other custom type.

`Feature` shoud expose a way to add instances of our custom extension, maybe using the same pattern as for other children (`AddChild` => `AddExtensionObject`). As far as I know, order can't be defined for extensions, so we don't have to worry about it.

If you have a clear idea of what you want to do but not enough time to do it, I can start the work following your instructions and submitting another merge request.

Again, thanks for your help !

**Comment by samcragg** [comment]
So I think there are a couple of options in this case.

One we would have two classes, `SimpleElement` and `ComplexElement` - both empty but you would inherit from them to add your custom data/attributes. Each class would then expose a collection of these, e.g.:

```C#
public class Feature
{
    public ICollection<SimpleElement> FolderSimpleExtensionGroup { get; }
    public ICollection<SimpleElement> FolderObjectExtensionGroup { get; }
}
```

The only thing with this is that the class will have multiple extension points, which is not a problem during serialization but during parsing we have to try and add them to the correct class.

An alternative might be to expose the `Orphans` property and allow any child to be added to an element. In your code you could then provide an extension methods on `Feature` to set/get this element to help tidy up your code:

```C#
// in main
KmlFactory.Register<MyOpacityElement>(...)

public static class FeatureExtensions
{
    public static MyOpacityElementGetOpacity(this Feature feature)
    {
        feature.Orphans.OfType<MyOpacityElement>().FirstOrDefault();
    }
}
```

**Comment by sylvaneau** [comment]
From my point of view, the first option seems to be better.

For me, the `Orphans` collection must remain read only. This collection is very handy as lets the developer know there are unknown (and potentially invalid) elements in the kml being parsed.

For parsing, indeed it's a bit more complicated, we might try to use the same pattern as the `ChildTypeAttribute`, but this time it will become something like `ParentTypeAttribute` :

```csharp
public class ParentTypeAttribute : Attribute
{
     public Type ParentType { get; }
     public int Order { get; }
}

[ParentType(typeof(Feature), 1)]
public class FeatureExtension : Element // or : ObjectElement see after
{
     ...
}
```

Then we have two options, using the `Children` collection or using a separated collection :

Option 1) When you call `KmlFactory.Register<FeatureExtension>` this will look for `ParentTypeAttribute` and then register this type as a child of `Feature`. So our extension will be added to the `Children` collection of the feature. Because technically, this extension is still a child of an element.

This works well for parsing, but `Children` is read only, so we cannot add a `FeatureExtension` object to our `Feature`. We need a way to add valid Child, a simple option would be to make `AddChild` public.

If we want to keep `AddChild` private, we can add a new method `AddObjectExtension(ObjectElement element)` to `Element` (or to any child type supporting extension).

```csharp
[ParentType(typeof(Feature), 1)]
public class FeatureExtension : ObjectElement
{
     ...
}

public class Element :
{
     public void AddObjectExtension(ObjectElement element)
     {
           this.AddChild<ObjectElement>(element);
     }
}
```

Order of elements is very important. A complex KML object will look like this :

```xml
<Placemark>
     <name>New CIP Organisation 3</name>
        <!-- FeatureExtension must be serialized after all elements inherited from Feature -->
     <ext:FeatureExtension>
        ...
     </ext:FeatureExtension>
     <!-- And before Placemark specific elements --> 
     <Point>
          <altitudeMode>clampToGround</altitudeMode>
          <coordinates>-33.380231099845957,-14.080241118090243,0</coordinates>
      </Point>
      <!-- Finaly, PlacemarkExtension must be after all Placemark specific elements -->
      <ext:PlacemarkExtension>
           ....
      </ext:PlacemarkExtension>
</Placemark>
```

When registering our `ParentTypeAttribute` we need to deal with the Order property with care. But I think it will work quite well as each type (even the abstract ones) have a list of valid child type with associated order.

Option 2) This option is entirely based on option 1, but we store extensions in separated collections instead of `Children`. I don't know if it's necesary, I prefer option 1 as Extensions could be considered as classic `Children` of a KML element. But I might miss something....

```csharp
public class Feature : ...
{
    public ICollection<SimpleElement> FeatureSimpleExtensionGroup { get; }
    public ICollection<ObjectElement> FeatureObjectExtensionGroup { get; }

    ...
}

public class Placemark : Feature
{
    public ICollection<SimpleElement> PlacemarkSimpleExtensionGroup { get; }
    public ICollection<ObjectElement> PlacemarkObjectExtensionGroup { get; }
}
```

**Comment by samcragg** [comment]
Giving this some thought and I like option 1 as well. So, to summarise the changes:

+ Add `ParentType` attribute, though I think the order may have to be the order you register them, as we won't be able to keep track of the order of extensions relative to each other, as they will all be added to the same collection as all the other valid child types. The alternative is the order refers to the absolute order relative to all children of that type, which I think will be confusing.
+ Upgrade `KmlFactory` to look for this attribute and add the class to the specified type. This will be added at the end of the existing valid children collection.
+ Expose `AddChild` and `Children`. Children will be changed to be stored in order, therefore, removing the need for `OrderedChildren`. Note we'll do the sorting (well, find out where to insert the item via [`BinarySearch`](https://docs.microsoft.com/en-gb/dotnet/api/system.collections.generic.list-1.binarysearch#System_Collections_Generic_List_1_BinarySearch__0_System_Collections_Generic_IComparer__0__)) each time a child is added rather than change the list to a [`SortedList`](https://docs.microsoft.com/en-gb/dotnet/api/system.collections.generic.sortedlist-2) due to memory concerns.

I think that's it. There's not too much work involved so I'll get that done this week but shout up if I've missed anything or it's something you'd like to have a go at submitting a PR for if it's urgent/you want to have a poke around the code yourself.

**Comment by samcragg** [comment]
That should be implemented this time then. I've updated the package and the documentation. Since the order of the registration is significant, I decided to go with an extra method on `KmlFactory` but there's no need to register the extension element and then register where it is attached to, you can just register it as an extension and it will get registered as a KmlElement if it's not been registered.

Have a quick look at [the docs](https://github.com/samcragg/sharpkml/blob/master/docs/Extending.md) and let me know if anything is unclear.

**Comment by sylvaneau** [comment]
Also, I've experienced a huge performance issue during file parsing. I think the problem is linked to the new `ElementCollection` (probably due to the sorting).

My test file takes more than a minute with the new code and about 5 seconds with the old one. I temporary made a revert of this commit on our branch.

You can find the test file joined to this comment.
[france_kml.zip](https://github.com/samcragg/sharpkml/files/2282961/france_kml.zip)

**Comment by samcragg** [comment]
Thanks for the example file. I can see the problem - with version 3.3.0 it takes 8.5 seconds to load on my machine but in 3.4.0 it takes 39 seconds. I'm currently benchmarking it and have it down to 4.5 seconds (the slowdown was due to the changes to the collection, however, I've taken the opportunity to do some profiling of the code in general)

I currently away from home so have limited time to look at it, but I am working on it and hopefully get something updated in the next week.

I'll also take a look at your fix; I appreciate your help and patience with extending the library so far

**Comment by sylvaneau** [comment]
Wonderfull !

I'll perform some tests right now and give you a feedback quickly !

Thanks again

**Comment by sylvaneau** [comment]
Good Job ! :-)

Everything seems to be OK :-) I just noticed a sligh memory increase compared to the previous version, but nothing to worry about.

Once again thanks a lot for your time !

I will come back soon with another merge request to improve file/memory management when dealing with large KMZ (and a LOOOT of attached files).

**Comment by samcragg** [comment]
This should be fixed now - your example has been merged into the Examples folder so hopefully everything should be OK now.
